src/Core.UnitTests/Notifications/NotificationServiceTests.cs

```diff
@@ -497,11 +497,44 @@ public void ShowNotification_UserClosedTheNotification_CriticalException_Excepti
             testSubject.ShowNotification(notification);
 
             Action act = () => infoBar.Raise(x => x.Closed += null, EventArgs.Empty);
-            act.Should().ThrowExactly<StackOverflowException>().And.Message.Should().Be("this is a test");
+            act.Should().ThrowExactly<StackOverflowException>().WithMessage("this is a test");
 
             logger.AssertNoOutputMessages();
         }
 
+        [TestMethod]
```

Comment:
These two tests are duplicates of the Tests "Dispose_NoExistingNotification_NoException" and "Dispose_HasExistingNotification_ExistingNotificationRemoved" because the dispose is just 
`Dispose(){RemoveExistingInfoBar()}`

and the new method is         `public void RemoveNotification() => RemoveExistingInfoBar();`

I am unsure how to handle this situation.

Reply:
Logically they are separate, so we should really test both.

The only difference between the tests is the method that is called (`Dispose` vs `RemoveNotification`). One way of removing the duplication is to move most of the test into a helper method that does the setup and checking, and accepts a delegate to the `Act` part of the test e.g.

```

[TestMethod]
public void RemoveNotification_HasACurrentNotification_RemovesCurrentNotification( )
=> Check_HasACurrentNotification_RemovesCurrentNotification(x => x.RemoveNotification());

[TestMethod]
public void Dispose_HasACurrentNotification_RemovesCurrentNotification( )
=> Check_HasACurrentNotification_RemovesCurrentNotification(x => x.Dispose());

        private void Check_HasACurrentNotification_RemovesCurrentNotification(Action<NotificationService> op)
        {
            var notification = CreateNotification();
            var infoBar = CreateInfoBar();
            var infoBarManager = CreateInfoBarManager(notification, infoBar.Object);
            var testSubject = CreateTestSubject(infoBarManager.Object);

            testSubject.ShowNotification(notification);

            VerifyInfoBarCreatedCorrectly(infoBarManager, notification);
            VerifySubscribedToInfoBarEvents(infoBar);

            // Call the product code
           op(testSubject);

            VerifyInfoBarRemoved(infoBarManager, infoBar);

            infoBarManager.VerifyNoOtherCalls();
        }

```

It's more complicated but removes the duplication. We've used it in a couple of places.

Reply:
I didnt know the syntax with (x => x.RemoveNotification()); and the action as a parameter existed, very cool